File: Java/com/leetcode/rotateList/RotateList.java
package com.leetcode.rotateList;

import com.leetcode.dataStructures.ListNode;

public class RotateList {
	public ListNode rotateRight(ListNode head, int n) {
        if (head == null || n == 0) {
            return head;
        }
        
        ListNode pt = head;
        ListNode tail = pt;
        int length = 0;
        while (pt != null) {
            length++;
            tail = pt;
            pt = pt.next;
        }
        if (n % length == 0) {
            return head;
        }
        
        int remained = length - (n % length);
        
        ListNode prev = null;
        pt = head;
        for (int i = 0; i < remained; i++) {
            prev = pt;
            pt = pt.next;
        }
        
        tail.next = head;
        prev.next = null;
        head = pt;
        return head;
    }
}

File: Python/rotate_list/rotate_list.py
class Solution(object):
    def rotateRight(self, head, k):
        if head is None:
            return head
        l = 0
        p = head
        while p:
            l += 1
            p = p.next
        k %= l
        if k == 0:
            return head
        i, p = l-k, head
        while i:
            i -= 1
            p = p.next
        q = p
        while q.next:
            q = q.next
        q.next = head
        i = l-k-1
        while i > 0:
            i -= 1
            head = head.next
        head.next = None
        return p
